libs/community/langchain_community/vectorstores/cratedb/base.py

```diff
@@ -0,0 +1,475 @@
+        )
+
+        if not connection_string:
+            raise ValueError(
+                "Database connection string is required."
+                "Either pass it as a parameter, or set the "
+                "CRATEDB_CONNECTION_STRING environment variable."
+            )
+
+        return connection_string
+
+    @classmethod
+    def connection_string_from_db_params(
+        cls,
+        driver: str,
+        host: str,
+        port: int,
+        database: str,
+        user: str,
+        password: str,
+    ) -> str:
+        """Return connection string from database parameters."""
+        return str(
+            sqlalchemy.URL.create(
+                drivername=driver,
+                host=host,
+                port=port,
+                username=user,
+                password=password,
+                query={"schema": database},
+            )
+        )
+
+    def _select_relevance_score_fn(self) -> Callable[[float], float]:
+        """
+        The 'correct' relevance function
+        may differ depending on a few things, including:
+        - the distance / similarity metric used by the VectorStore
+        - the scale of your embeddings (OpenAI's are unit normed. Many others are not!)
+        - embedding dimensionality
+        - etc.
+        """
+        if self.override_relevance_score_fn is not None:
+            return self.override_relevance_score_fn
+
+        # Default strategy is to rely on distance strategy provided
+        # in vectorstore constructor
+        if self._distance_strategy == DistanceStrategy.COSINE:
+            return self._cosine_relevance_score_fn
+        elif self._distance_strategy == DistanceStrategy.EUCLIDEAN:
+            return self._euclidean_relevance_score_fn
+        elif self._distance_strategy == DistanceStrategy.MAX_INNER_PRODUCT:
+            return self._max_inner_product_relevance_score_fn
+        else:
+            raise ValueError(
+                "No supported normalization function for distance_strategy of "
+                f"{self._distance_strategy}. Consider providing relevance_score_fn to "
+                "CrateDBVectorStore constructor."
+            )
+
+    @staticmethod
+    def _euclidean_relevance_score_fn(similarity: float) -> float:
+        """Return a similarity score on a scale [0, 1]."""
+        # The 'correct' relevance function
+        # may differ depending on a few things, including:
+        # - the distance / similarity metric used by the VectorStore
+        # - the scale of your embeddings (OpenAI's are unit normed. Many
+        #  others are not!)
+        # - embedding dimensionality
+        # - etc.
+        # This function converts the euclidean norm of normalized embeddings
+        # (0 is most similar, sqrt(2) most dissimilar)
+        # to a similarity function (0 to 1)
+
+        # Original:
+        # return 1.0 - distance / math.sqrt(2)
+        return similarity / math.sqrt(2)
```

Comment:
Hi @BaurzhanSakhariev. Thanks for your reply above. I am all in for better naming things. Compared to the original variant still visible on the PR, the argument name already had been renamed to `similarity`, as I assumed this is the kind/unit of value that will be emitted from CrateDB's `vector_similarity()` function.

Please advise if it should be renamed once more, like ... as you suggested.

Reply:
Ah, so similarity argument here is a value returned from CrateDB?

Then we don't need anything on top of it, ie we don't need this function and instead should use value from CrateDB as is.

Similarity is already in the (0,1] interval and dividing by math.sqrt(2) won't normalize it but return wrong result, for example 1 will become  0.714.

Reply:
Thank you very much. The patch has been adjusted with 7ff13192b1f.

Reply:
Thanks!

Just checking the commit, looks like we had exactly one instance of
>for example 1 will become 0.714.

```
-(Document(page_content="foo", metadata={"page": "0"}), 0.7071067811865475),
+(Document(page_content="foo", metadata={"page": "0"}), 1.0),
```


Checked test, it must be 1 indeed => we have same value -> vectors coincide -> return 1 which is max similarity.
docsearch.similarity_search_with_relevance_scores("foo", k=3) and page_content="foo" are same

Reply:
> looks like we had exactly one instance of "1 will become 0.714"

Exactly, your admonition was right on the spot. Thanks again!